File: solution1.py
# https://leetcode.com/problems/coin-change/

# // Time Complexity : o(row*col) , row = number of coins + 1, col = amount+1
# // Space Complexity : o(row*col) , row = number of coins + 1, col = amount+1
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this : I got so lost in the indexing....
#
#
# // Your code here along with comments explaining your approach
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        MAX = 99999
        coins.insert(0, 0)
        row, col = len(coins), amount + 1
        dp = [[None for i in range(col)] for j in range(row)]

        # initialize
        for i in range(col):
            dp[0][i] = MAX
        for i in range(row):
            dp[i][0] = 0

        for i in range(1, row):
            for j in range(1, col):
                if j < coins[i]:
                    dp[i][j] = dp[i - 1][j]
                else:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - coins[i]] + 1)
        if dp[row - 1][col - 1] == MAX:
            return -1
        else:
            return dp[row - 1][col - 1]

File: Solution1.java
// Leetcode Problem : https://leetcode.com/problems/coin-change/description/
// Time Complexity : O(n*m) where n is the total number of coins and m is the total amount
// Space Complexity : O(m) where m is the total amount
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
 * I create a single dimensional dp[amount+1] array and store all the values as amount+1 
 * initially. I iterate over the dp array as many times as there are total number of coins.
 * If coins[i] <= j I set dp[j] as the minimum of dp[j] and 1 + dp[j-coins[i]].
 * If the last element of dp array is still equal to amount+1 I return -1 else I'll return the last element.
 */
class Solution {
    public int coinChange(int[] coins, int amount) {
        int dp[] = new int[amount+1];
        for(int i=1;i<dp.length;i++)
        {
            dp[i]=amount+1;
        }
        for(int i=0;i<coins.length;i++)
        {
            for(int j=1;j<dp.length;j++)
            {
                if(coins[i]<=j)
                {
                    dp[j] = Math.min(dp[j],1+dp[j-coins[i]]);
                }
            }
        }
        if(dp[amount]>amount)
        {
            return -1;
        }
        return dp[amount];
    }
}